packages/broadcast/src/common/broadcast-client.ts

```diff
@@ -0,0 +1,55 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import { IBroadcastProtocol, IBroadcastServer, IBroadcastState } from './broadcast-protocol';
+import { BroadcastWatcher } from './broadcast-watcher';
+
+export const IBroadcastClientDispatch = Symbol('IBroadcastClientDispatch');
+
+type onStateUpdateHandler = (event: IBroadcastProtocol) => void;
+
+export interface IBroadcastClientDispatch {
+  onStateUpdate(callback: onStateUpdateHandler): void;
+  getState(): Promise<IBroadcastState>;
+  setState(state: IBroadcastState): Promise<IBroadcastProtocol>;
+}
+
+@injectable()
+export class BroadcastClient implements IBroadcastClientDispatch {
+
+  constructor(
+    @inject(BroadcastWatcher) private readonly watcher: BroadcastWatcher,
+    @inject(IBroadcastServer) private readonly server: IBroadcastServer
+  ) {}
+
+  onStateUpdate(callback: onStateUpdateHandler) {
+    this.watcher.onStateUpdate(callback);
+  }
+
+  getState(): Promise<IBroadcastState> {
+    return this.server.getState();
+  }
+
+  setState(state: IBroadcastState): Promise<IBroadcastProtocol> {
+    const prevState = this.server.getState();
+    return new Promise(() => {
+      prevState.then(() => this.server.setState(state));
```

Comment:
I don't understand how TypeScript understands this and lets it pass: to me this `setState` function actually returns a `Promise<void>`, am I missing something?

I think a better implementation in this case would be something like the following:

```ts
    async setState(state: IBroadcastState): Promise<IBroadcastProtocol> {
        await this.server.getState();
        return this.server.setState(state);
    }
```

wdyt?

Reply:
Looks better, thanks. Doing it right away!